File: site-toolkit/components/core/src/main/java/org/hippoecm/hst/core/util/DOMUtils.java
/*
 *  Copyright 2008-2013 Hippo B.V. (http://www.onehippo.com)
 * 
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 * 
 *       http://www.apache.org/licenses/LICENSE-2.0
 * 
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.hippoecm.hst.core.util;

import java.io.IOException;
import java.io.StringWriter;

import org.hippoecm.hst.util.DOMElementWriter;
import org.w3c.dom.Element;

/**
 * DOMUtils
 * 
 * @version $Id$
 */
public class DOMUtils {

    public static final int DEFAULT_ELEMENT_STRINGIFYING_BUFFER_SIZE = 80;
    public static final int DEFAULT_INDENT = 0;
    public static final String DEFAULT_INDENT_WITH = "\t";

    private DOMUtils() {
    }

    public static String stringifyElement(Element element) {
        return stringifyElement(element, DEFAULT_ELEMENT_STRINGIFYING_BUFFER_SIZE, DEFAULT_INDENT, DEFAULT_INDENT_WITH);
    }

    public static String stringifyElement(Element element, int initialBufferSize, int indent, String indentWith) {
        String stringified = null;
        StringWriter writer = new StringWriter(initialBufferSize);

        try {
            DOMElementWriter domWriter = new DOMElementWriter();
            domWriter.write(element, writer, indent, indentWith);
        } catch (IOException e) {
        }

        stringified = writer.toString();
        return stringified;
    }

    public static String stringifyElementToHtml(Element element) {
        return stringifyElement(element);
    }

    public static String getIdAttribute(Element element) {
        String value = null;

        if (element != null) {
            if (element.hasAttribute("id"))
                value = element.getAttribute("id");
            else if (element.hasAttribute("ID"))
                value = element.getAttribute("ID");
            else if (element.hasAttribute("Id"))
                value = element.getAttribute("Id");
            else if (element.hasAttribute("iD"))
                value = element.getAttribute("iD");
        }

        return value;
    }

}

File: channel-manager/frontend-ng/src/app/iframe/utils/dom.utils.js
/*
 * Copyright 2022 Bloomreach. All rights reserved. (https://www.bloomreach.com/)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

async function loadImage(image, onLazyLoadCallback) {

  const loadingAttribute = image.getAttribute('loading');
  const isLazy = loadingAttribute && loadingAttribute.trim().toLowerCase() === 'lazy';

  return window.$Promise((resolve, reject) => {

    function fulfill() {
      if (isLazy) {
        onLazyLoadCallback(image);
      } else if (image.naturalWidth) {
        resolve(image);
      } else {
        reject(image);
      }

      image.removeEventListener('load', fulfill);
      image.removeEventListener('error', fulfill);
    }

    if (image.naturalWidth) {
      // If the browser can determine the naturalWidth the image is already loaded successfully
      resolve(image);
    } else if (image.complete) {
      // If the image is complete but the naturalWidth is 0px it is probably broken
      reject(image);
    } else {
      image.addEventListener('load', fulfill);
      image.addEventListener('error', fulfill);

      if (isLazy) {
        // resolve the main promise for lazy-load images immediately
        resolve(image);
      }
    }
  });
}

async function loadImages(input, onLazyLoadCallback) {
  if (input.length === undefined || input.length === 0) {
    // if there are no images to wait for, we immediately resolve
    return window.$Promise.resolve();
  }

  // otherwise, wait for each image to load completely before resolving
  // errors are caught and stored in the promise result
  return window.$Promise.all(input.map((img) => loadImage(img, onLazyLoadCallback).catch((error) => error)));
}


/**
 * Wait for all image elements in the "images" array to finish loading before calling "onLoadCallback".
 *
 * If an image has attribute loading="lazy", the "onLazyLoadCallback" will be used to issue an async callback. This
 * means that the callback is invoked once the browser determines that the image should be loaded (e.g. when the user
 * scrolls). This does introduce a bit of overhead for lazy-load images that are already visible to the user as it will
 * trigger multiple callbacks.
 *
 * @param {*} images The images to wait for
 * @param {*} onLoadCallback  Callback to invoke once all images have finished loading
 * @param {*} onLazyLoadCallback  Callback to invoke once a "lazy-load" image has finished loading
 */
export async function waitForImagesToLoad(images, onLoadCallback, onLazyLoadCallback = () => {}) {
  try {
    await loadImages(images.get(), onLazyLoadCallback);
  } finally {
    onLoadCallback();
  }

}
